Pluggable state backend support for checkpoints

gyfora opened the pull request:
This PR introduces the pluggable state backends using StateHandleProviders and extends the StateHandle interface with a discard method for cleaning up the unnecessary checkpoints.

It also adds a statehandle/provider implementation for storing checkpoints in any flink supported file system such as HDFS or Tachyon.

The checkpoint coordinator has been modified to properly discard user state handles using the following logic:
 - If a pending checkpoint expires (by the timer thread) it discards the user state
 - When a successful checkpoint expires (by acquiring following successful ones) it discards user states
 - When the checkpoint coordinator is shut down it discards all pending and successful states

Travis error:
I modified the recovery IT case to use the local file system for storing the checkpoints. Afterwards it checks whether the directory is empty. The test passes if I only run that specific test locally , but it seems to fail when I run with the other IT cases for no apparent reason. Usually a couple of files (2-5) remain in the checkpoint directory, meaning that almost all of them had been deleted but those. Also the checkpointing and recovery logic runs fine without test failure.

Could it happen that the jobmanager doesnt shut down the checkpointmanager in time?

I would appreciate some help figuring this out somehow, or trying to reproduce it locally.

mbalassi commented:
Changes look good to me. Some documentation should be added on the behaviour and maybe a iscussion on the dev list would be also nice on how to proceed, like incremental checkpointing and the general architecture of the solution.

If you have the time I'm happy to give you a helping hand with the test over the weekend.

gyfora commented:
I commented out the check in the recovery test that caused the issue until I figure it out. It's definitely not a blocker as it seems that only the very last successful checkpoints remain in the directory when it runs together with other tests.

mbalassi commented:
As far as I see the problem is that the `AbstractProcessFailureRecoveryTest` that you use as a base class only makes sure that the anything that is in the `programTrigger` thread is actually finished. As you are running the `CheckpointCoordinator` on the JobManager side there is absolutely no guarantee for that part of the code in terms of termination.

gyfora commented:
I added a test for the FileStateHandle with HDFS, and I also added a commit that fixes an unrelated (but important) issue to the end. 

I would like to merge this soon, so we can build on it afterwards.

@StephanEwen @rmetzger could you guys take a quick look? It's straightforward I guess...

senorcarbone commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/ByteStreamStateHandle.java:
```diff
@@ -0,0 +1,77 @@
+
+	private static final long serialVersionUID = -962025800339325828L;
+
+	private transient Serializable state;
```
The state here might potentially persist in the StateHandle as is (and get transferred to the JobManager) apart from being written to disk, if it happens to be in the same JVM. There is not guaranteed Serialisation with Akka...

gyfora replied:
We manually serialize all the time in the RuntimeEnvironment acknowledgeCheckpoint method into a StateForTask object :)
but good catch

senorcarbone replied:
oh found it, indeed. It's the SerializedValue constructor actually. That's nice, good job

senorcarbone commented:
Apart from my comment on the transient field I think this is ok to merge. Perhaps we can look into standardising the task manager backend configuration afterwards. So it's :+1: from me :)
